### PR TITLE
fix!: Integer division is not the inverse of integer multiplication

### PR DESCRIPTION
**Pull request by jfecher**
# Description

## Problem\*

Resolves https://github.com/noir-lang/noir/issues/6242

## Summary\*

Making this PR for visibility to show that fixing this issue breaks our serialize test.

Notably this fix allows rounding arithmetic generics using the `/ N * N` pattern which was simplified away previously.

## Additional Context

I initially removed the cancellation of / and * entirely but found that this breaks even more tests. So I added `approx_inverse` in a few cases that only involve constants that I thought may be correct but we still fail `serialize` even with these cases.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: compiler/noirc_frontend/src/hir_def/types/arithmetic.rs [hunk 1]
```diff
@@ -125,7 +125,7 @@ impl Type {
 
         // Note that this is exact, syntactic equality, not unification.
         // `rhs` is expected to already be in canonical form.
-        if l_op.inverse() != Some(op) || l_rhs.canonicalize() != *rhs {
+        if l_op.approx_inverse() != Some(op) || l_rhs.canonicalize() != *rhs {
```
Comment (michaeljklein) [inline compiler/noirc_frontend/src/hir_def/types/arithmetic.rs:128]: Should we emit a warning in the cases where `approx_inverse` is used?

**Review by michaeljklein — CHANGES_REQUESTED** [review]
Question about warning r.e. `approx_inverse`, otherwise LGTM

**Comment by jfecher** [comment]
> Question about warning r.e. approx_inverse, otherwise LGTM

Even with `approx_inverse` we're still failing to compile the serialize test. It seems that we rely on this simplification that this PR removes so I'm not sure on a path forward.

**Review by michaeljklein — APPROVED** [review]
LGTM

File: compiler/noirc_frontend/src/hir_def/types/arithmetic.rs (hunk 1 above)
Comment (jfecher) [reply compiler/noirc_frontend/src/hir_def/types/arithmetic.rs]: No, I don't think a warning is needed. We should limit uses of approx_inverse to cases where we know it is correct, e.g. when we're checking if something divides evenly

**Pull request merged**